### PR TITLE
[WasmFS] Add a hack to work around deadlocks

### PR DESCRIPTION
**Pull request by tlively**
When the main thread tries to perform a WasmFS operation that tries to acquire a
lock that is already held by another thread that is trying to proxy a backend
operation, deadlock can occur. Proxying depends on passing a message to a worker
thread and that message has to be relayed through the main thread. When the main
thread is spinning on a lock, the proxying message is not relayed and neither
thread makes forward progress.

Work around the issue by adding a heartbeat event on the dedicated worker thread
to make it execute the queue periodically even without receiving a message from
the main thread. This allows the proxying thread holding the lock to make
forward progress and resolves the deadlock.

### CODE REVIEW

**Comment by tlively** [comment]
Current dependencies on/for this PR:
* main
  * **PR #17869** <a href="https://app.graphite.dev/github/pr/emscripten-core/emscripten/17869" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/emscripten-core/emscripten/17869?utm_source=stack-comment).

File: system/lib/wasmfs/thread_utils.h [hunk 1]
```diff
@@ -25,6 +25,24 @@ class ProxyWorker {
...
+      // if the main thread is blocked.
+      //
+      // Note that this requires adding _emscripten_proxy_execute_queue to
+      // EXPORTED_FUNCTIONS.
+      EM_ASM({
+          var heartbeat = () => {
+            _emscripten_proxy_execute_queue($0);
+            setTimeout(heartbeat, 50);
+          };
+          heartbeat();
```
Comment (kripken) [inline system/lib/wasmfs/thread_utils.h:43]:
```suggestion
          var intervalID = setInterval(() => {
            if (ABORT) {
              clearInterval(intervalID);
            } else {
              _emscripten_proxy_execute_queue($0);
            }
          }, 1000);
```

File: system/lib/wasmfs/thread_utils.h [hunk 2]
```diff
@@ -25,6 +25,24 @@ class ProxyWorker {
...
   ProxyWorker()
     : queue(), thread([&]() {
+
+      // Sometimes the main thread is spinning, waiting on a WasmFS lock held by
+      // a thread trying to proxy work to this dedicated worker. In that case,
+      // the proxying message won't be relayed by the main thread and the system
+      // will deadlock. This heartbeat ensures that proxying work eventually
+      // gets done so the thread holding the lock can make forward progress even
+      // if the main thread is blocked.
+      //
```
Comment (kripken) [inline system/lib/wasmfs/thread_utils.h:35]:
```suggestion
      // TODO: remove this once we can postMessage directly from any worker to another, as then the main thread will not be involved
```

File: system/lib/wasmfs/thread_utils.h [hunk 3]
```diff
@@ -25,6 +25,29 @@ class ProxyWorker {
...
+        var intervalID = setInterval(() => {
+          if (ABORT) {
+            clearInterval(intervalID);
+          } else {
+            _emscripten_proxy_execute_queue($0);
+          }
+        }, 50);
+      }, queue.queue);
+
         // Sit in the event loop performing work as it comes in.
```
Comment (kripken) [inline system/lib/wasmfs/thread_utils.h:51]: This indentation looks wrong. Should the entire EM_ASM be indented two more?

**Review by kripken — APPROVED** [review]

**Pull request merged**